src/storage/transaction/ChainAddEdgeProcessor.cpp

```diff
@@ -0,0 +1,139 @@
+        });
+    LOG_IF(INFO, FLAGS_trace_toss) << "-prepareLocal() tid=" << txnId_;
+    return std::move(c.second);
+}
+
+
+folly::SemiFuture<cpp2::ErrorCode> ChainAddEdgesProcessor::processRemote(cpp2::ErrorCode code) {
+    LOG_IF(INFO, FLAGS_trace_toss) << "processRemote() " << txnId_;
+    setErrorCode(code);
+    if (code_ != cpp2::ErrorCode::SUCCEEDED) {
+        return code_;
+    }
+    std::vector<cpp2::NewEdge> outEdges(inEdges_.size());
+    std::transform(inEdges_.begin(), inEdges_.end(), outEdges.begin(), [&](auto edge) {
+        std::swap(edge.key.src, edge.key.dst);
+        edge.key.edge_type = 0 - edge.key.edge_type;
+        if (spaceVidType_ == meta::cpp2::PropertyType::INT64) {
+            TransactionUtils::changeToIntVid(edge.key);
+        }
+        return edge;
+    });
+
+    auto c = folly::makePromiseContract<cpp2::ErrorCode>();
+    doRpc(outEdges, std::move(c.first));
+    LOG_IF(INFO, FLAGS_trace_toss) << "-processRemote() " << txnId_;
+    return std::move(c.second);
+}
+
+
+/**
+ * @brief call sClient::addEdges, if leader changed, retry
+ */
+void ChainAddEdgesProcessor::doRpc(std::vector<cpp2::NewEdge>& outEdges,
+                                   folly::Promise<cpp2::ErrorCode>&& pro) noexcept {
+    LOG_IF(INFO, FLAGS_trace_toss) << "doRpc() " << txnId_;
+    auto* sClient = env_->txnMan_->getStorageClient();
+    sClient->addEdges(spaceId_, outEdges, propNames_, overwritable_)
+        .via(env_->txnMan_->getExecutor())
+        .thenTry([=, p = std::move(pro)](auto&& ret) mutable {
+            auto rc = cpp2::ErrorCode::SUCCEEDED;
+            if (ret.hasValue()) {
+                StorageRpcResponse<cpp2::ExecResponse>& rpcResp = ret.value();
+                if (!rpcResp.failedParts().empty()) {
+                    CHECK_EQ(rpcResp.failedParts().size(), 1);
+                    rc = rpcResp.failedParts().begin()->second;
+                    if (rc == cpp2::ErrorCode::E_LEADER_CHANGED) {
+                        LOG_IF(INFO, FLAGS_trace_toss) << "doRpc() leader changed, retry";
+                        doRpc(outEdges, std::move(p));
+                        return;
+                    }
+                }
+            } else {
+                rc = cpp2::ErrorCode::E_FORWARD_REQUEST_ERR;
+            }
+            LOG_IF(INFO, FLAGS_trace_toss)
+                << "~doRpc(), tid=" << txnId_ << ", rc=" << CommonUtils::name(rc);
+            p.setValue(rc);
+            return;
+        });
+}
+
+folly::SemiFuture<cpp2::ErrorCode> ChainAddEdgesProcessor::processLocal(cpp2::ErrorCode code) {
+    LOG_IF(INFO, FLAGS_trace_toss) << "processLocal() " << txnId_;
+    setErrorCode(code);
+    if (code_ != cpp2::ErrorCode::SUCCEEDED) {
+        return code_;
+    }
+
+    // 1. remove locks
+    kvstore::BatchHolder bat;
+    std::for_each(inEdges_.begin(), inEdges_.end(), [&](auto& e) {
+        bat.remove(TransactionUtils::lockKey(vIdLen_, partId_, e.key));
+        auto ret = encoder_(e);
```

Comment:
AddEdgesAtomicProcessor::encodeEdge not handle` std::make_pair("", cpp2::ErrorCode::E_SPACE_NOT_FOUND);`